pkg/qemu/devices.go

```diff
@@ -81,20 +88,115 @@ func (rod ReadOnlyDirectory) Cmdline() []string {
 		"-device", "ide-drive,drive=tmpdir,bus=ahci.0",
 	}
 }
+func (ReadOnlyDirectory) KArgs() []string { return nil }
+func (rod ReadOnlyDirectory) Cleanup(tb testing.TB) {
+	if tb.Failed() {
+		tb.Logf("keeping temp dir: %s", rod.Dir)
+	} else {
+		if err := os.RemoveAll(rod.Dir); err != nil {
+			tb.Logf("failed to remove temporary directory %s: %v", rod.Dir, err)
+		}
+	}
+}
+
+// P9Directory is a Device that exposes a directory as a Plan9 (9p)
+// read-write filesystem in the VM.
+type P9Directory struct {
+	// Dir is the directory to expose as read-write 9p filesystem.
+	Dir string
+
+	// Boot: if true, indicates this is the root volume. For this to work,
+	// kernel args will need to be added - use KArgs() to get the args. There
+	// can only be one boot 9pfs at a time.
+	Boot bool
+
+	// Tag is an identifier that is used within the VM when mounting an fs,
+	// e.g. 'mount -t 9p my-vol-ident mountpoint'. If not specified, a default
+	// tag of 'tmpdir' will be used.
+	//
+	// Ignored if Boot is true, as the tag in that case is special.
+	//
+	// For non-boot devices, this is also used as the id linking the `-fsdev`
+	// and `-device` args together.
+	//
+	// Because the tag must be unique for each dir, if multiple non-boot
+	// P9Directory's are used, tag may be omitted for no more than one.
+	Tag string
+}
+
+func (p P9Directory) Cmdline() []string {
+	if len(p.Dir) == 0 {
+		return nil
+	}
+
+	var tag, id string
+	if p.Boot {
+		tag = "/dev/root"
+	} else {
+		tag = p.Tag
+		if len(tag) == 0 {
+			tag = "tmpdir"
+		}
+	}
+	if p.Boot {
+		id = "rootdrv"
+	} else {
+		id = tag
+	}
+
+	// Expose the temp directory to QEMU
+	return []string{
+		// security_model=mapped-file seems to be the best choice. It gives
+		// us control over uid/gid/mode seen in the guest, without requiring
+		// elevated perms on the host.
+		"-fsdev", fmt.Sprintf("local,id=%s,path=%s,security_model=mapped-file", id, p.Dir),
+		"-device", fmt.Sprintf("virtio-9p-pci,fsdev=%s,mount_tag=%s", id, tag),
+	}
+}
+func (p P9Directory) KArgs() []string {
```

Comment:
did you gofmt this file?

Reply:
Pretty sure I had, and it doesn't make any changes in that file when I run it now.

Reply:
Weird. I thought gofmt enforced an empty line between each function.

Reply:
yeah I was thinking it did between multi-line functions :man_shrugging: 
I can insert newlines if you like.

(fwiw, running `go fmt .` in u-root/ resulted in a bunch of changes in elvish and a few other cmd's)

Reply:
It looks like the circleci test for gofmt broke somehow. I'll push a quick fix.